Sources/Redis/Client/RedisClient+Pipeline.swift

```diff
@@ -0,0 +1,59 @@
+import Foundation
+
+public final class RedisPipeline {
+    /// The commands in this pipeline
+    var commands = [RedisData]()
+    
+    /// Adds a command to the pipeline
+    public func command(_ command: String, args: [RedisData] = []) {
+        let command = RedisData.array([RedisData(bulk: command)] + args)
+        self.commands.append(command)
+    }
+    
+    /// Adds a command to the pipeline
+    public func command(_ command: String, args: RedisData...) {
+        self.command(command, args: args)
+    }
+}
+
+extension RedisClient {
+    /// Executes a pipelined transaction
+    public func pipeline(closure: (RedisPipeline)->()) -> Future<[RedisData]> {
```

Comment:
I'm not sure this is the best API to get back the pipelined data. I'd prefer something like:

```swift
let a = redis.get("foo") // ELFuture<RedisData>
let b = redis.get("bar") // ELFuture<RedisData>
a.and(b).map {
    let (a, b) = $0
    print(a) // RedisData
    print(b) // RedisData
}
```

I know this doesn't work ATM because of the `currentSend` restrictions, but that issue will be resolved in the next version of Redis. Because of that, I don't really want to add new API that won't be necessary next release.

Reply:
Got ya.  I totally understand if you are about to rewrite something, you don't want to create a new API.  So no worries with not wanting to merge this.

However, I still question the API that is proposed.  For a Redis pipeline, you have to do the following from the client perspective

 -> SUBMIT -> GET foo
 -> SUBMIT -> GET bar
 -> WAIT -> GET foo
 -> WAIT -> GET bar

There is no way for your first call to "GET foo" to know that it needs to hang on and submit "GET bar" before performing the "WAIT foo".  The code you proposed will still perform the 2 transactions but not as a pipeline.  It is just lining them up as follows

 -> SUBMIT -> GET foo
 -> WAIT -> GET foo
 -> SUBMIT -> GET bar
 -> WAIT -> GET bar

and using a future to give the impression that it was a transaction, but they still performed 2 round trips to the server.

Anyways, I just don't see a way for "GET foo" to know that more transaction are coming BEFORE inserting the WAIT that it would normally do by default.

Reply:
Honestly, the ideal API would be if all of the helper methods were implemented in a protocol except the "command" method, and then that was implemented by the client and by a proxy.  Simple example below (ignore any syntax issues)

```swift
public protocol RedisClientProtocol {
    func get(_ key: String) -> Future<RedisData>
    func command(_ command: String, _ args: [RedisData]) -> Future<RedisData>
}

extension RedisClientProtocol {
    func get(_ key: String) -> Future<RedisData> {
        return self.command(key, [])
   }
}

public final class RedisClient: RedisClientProtocol {
    func command(_ command: String, _ args: [RedisData]) -> Future<RedisData> {
        // Send the command
    }
}

public final class RedisTransaction: RedisClientProtocol {
    func command(_ command: String, _ args: [RedisData]) -> Future<RedisData> {
        // Add the command to a pipeline of commands that will later be sent together using the client
    }
}
```

Basically this would give you all of the helper methods inside of the pipeline closure, since the helper methods are expecting a protocol rather than a specific object.

The API I came up with was the best compromise.  I lost the helpers but the transactions are properly lined up such that they are sent all together and then start waiting for responses.

Reply:
> There is no way for your first call to "GET foo" to know that it needs to hang on and submit "GET bar" before performing the "WAIT foo". 

That's actually exactly what this code snippet does:

```swift
let a = redis.get("foo") // ELFuture<RedisData>
let b = redis.get("bar") // ELFuture<RedisData>
a.and(b).map {
    let (a, b) = $0
    print(a) // RedisData
    print(b) // RedisData
}
```

Both of the calls to `get` return an `EventLoopFuture`. That means no waiting for any response has happened yet. The waiting happens once you call `map` on the combined futures.

The above snippet would yield the following:

```
-> SUBMIT -> GET foo
-> SUBMIT -> GET bar
-> WAIT -> GET foo
-> WAIT -> GET bar
```

Reply:
Thanks @tanner0101.

I think my confusion is that I don't think that code will do what you stated.  I agree both requests will generate a future, and I agree the requests will be back to back, but each request will actually request a round trip to the server.

Now, the piece I might be missing is that in the rework, you might be updating it so that all calls that are consecutive in the same "run of the run loop" actually default to pipeling as described here https://redis.io/topics/pipelining

If that is the case, then I absolutely agree that we don't need to do anything special since you are defaulting it, and we can just use futures.

I think my confusion is that currently, you have to have ALL of the commands available when you call the "send" method in order for it to pipeline (even if you comment out "currentSend") meaning you do actually have to have a special API to activate the transaction.

Reply:
> all calls that are consecutive in the same "run of the run loop" actually default to pipeling 

How the data is actually sent over the wire is entirely up to NIO, and I don't think we need to worry about that here. Whether or not we combine the entire pipelined request into a single ByteBuffer shouldn't matter. Each ByteBuffer we emit synchronously will be written to the OS socket and sent to the connected server as quickly as possible. 

> but each request will actually request a round trip to the server.

That will not be the case for the snippet I wrote. Each call to `get` will _synchronously_ enqueue the data to be sent to the server. If the socket is writable, that requests will be sent immediately. At no point here are we waiting for a response from the server to send the next request.

Once we have synchronously sent the two requests, the next part of the code with `map` then waits for both of the responses from the server. These responses are expected back in order, based on the order they were sent.

Reply:
Got ya @tanner0101.  I was looking at the deep NIO code and I thought passing it an array of values on the QueryHandler "enqueue" command would actually line it up to do it on a single continuous transaction, but sounds like that is not the case.

Is NIO just an open socket to the server for transmits and an open socket coming back for responses?  If that is the case, then it sounds like there is truly "no control" but also that means you probably don't care either because you are never truly waiting.  You just keep shoveling it commands and it keeps shoveling responses back vs. the typical "request/response" model.

If this is the case, then I totally agree with what you are saying, and that being said, we don't need any sort of API for pipelining.  Then I would also agree that the "multi/exec" should just be some helpers.

Reply:
@ericchapman `QueueHandler` does actually work that way, you're right. But that is not a part of NIO, that's a part of vapor/core. We're deprecating that in Vapor 4, and not using it in the next version of Redis. `QueueHandler` is the reason we need `currentSend` and really the main thing blocking us from supporting true pipelining. 

> Is NIO just an open socket to the server for transmits and an open socket coming back for responses?

Yes. By default it will write and read data as soon as the OS allows it. But again I don't think we need to worry about things at that low of a level here. Redis defines pipelining as:

> A Request/Response server can be implemented so that it is able to process new requests even if the client didn't already read the old responses.

This is not true of the current version of this package, because of QueueHandler and `currentSend` restrictions. However, this is definitely true of the current development version of Redis. You can send as many requests as you want without having to read old responses. When you decide to start waiting for the responses, they will be fulfilled in the order they were sent. This is true pipelining.

Reply:
Thanks @tanner0101.  I think we are now aligned and thank you for your patience in helping me understand this :)

So looks like the rework you are working should address all of my concerns and make it so everything "just works".  I do agree we shouldn't create any new APIs for this given they won't be necessary later.

I think for the short term, I'll just have to use my created Fork so I can work around these issues, then I can port my code over whenever the next version is finished.

Do we just reject this PR and close it?  Also, ping me if you need any help with the next version.  I am HIGHLY motivated to make this library extremely performant and error free ;).

Reply:
Yeah I think let's close this one for now. I still think a convenience method for using `MULTI` would be helpful, but that can be a separate PR. 

Help on the next version of Redis would definitely be appreciated! Here are links to the various repos:
https://github.com/mordil/nio-redis
- RedisConnection type. Channel handlers for encoding/decoding RESP protocol.
- Mostly done, but any help testing performance would be greatly appreciated.
https://github.com/vapor/redis-kit
- Connection pooling. Convenience methods for various commands should go here.
- Needs a lot of help. Currently missing almost all of the convenience methods. 
https://github.com/vapor/redis/pull/139
- Combines RedisKit + Vapor
- Should be very light, what's in that PR should be most of what we need